### PR TITLE
Wrong biome list for 1.19.4 if datapacks are not enabled

### LINKED ISSUE #4015: [v7] DEV BUILD 7.0 error creating world
### Server Implementation

Paper

### Server Version

1.19.3

### Describe the bug

[00:02:03 ERROR]: [Multiverse-Core] The world 'CityBuild' could NOT be loaded because it contains errors and is probably corrupt!
[00:02:03 ERROR]: [Multiverse-Core] Try using Minecraft Region Fixer to repair your world! 'CityBuild'
[00:02:03 ERROR]: [Multiverse-Core] https://github.com/Fenixin/Minecraft-Region-Fixer
[00:02:03 ERROR]: [Multiverse-Core] Failed to Create/Load the world 'CityBuild'
[00:02:03 INFO]: CONSOLE: FAILED.

### To Reproduce

plot setup (finish all and set name then error)

### Expected behaviour

[00:02:03 ERROR]: [Multiverse-Core] The world 'CityBuild' could NOT be loaded because it contains errors and is probably corrupt!
[00:02:03 ERROR]: [Multiverse-Core] Try using Minecraft Region Fixer to repair your world! 'CityBuild'
[00:02:03 ERROR]: [Multiverse-Core] https://github.com/Fenixin/Minecraft-Region-Fixer
[00:02:03 ERROR]: [Multiverse-Core] Failed to Create/Load the world 'CityBuild'
[00:02:03 INFO]: CONSOLE: FAILED.

### Screenshots / Videos

_No response_

### Error log (if applicable)

_No response_

### Plot Debugpaste

https://athion.net/ISPaster/paste/view/48fdb008bd06469e95f1421bc7ca585c

### PlotSquared Version

7.0 DEV BUILD (git repo download and build)

### Checklist

- [X] I have included a Plot debugpaste.
- [X] I am using the newest build from https://www.spigotmc.org/resources/77506/ and the issue still persists.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by TheMeinerLP**
## Overview

Fixes #4015

## Description
This is a workaround for use PSv7 without cherry grove biome

## Submitter Checklist

- [X] Make sure you are opening from a topic branch (**/feature/fix/docs/ branch** (right side)) and not your main branch.
- [X] Ensure that the pull request title represents the desired changelog entry.
- [X] New public fields and methods are annotated with `@since TODO`.
- [X] I read and followed the [contribution guidelines](https://github.com/IntellectualSites/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java [hunk 1]
```diff
@@ -436,6 +438,9 @@ private final class BukkitPlotBiomeProvider extends BiomeProvider {
         static {
             ArrayList<Biome> biomes = new ArrayList<>(List.of(Biome.values()));
             biomes.remove(Biome.CUSTOM);
+            if (PaperLib.getMinecraftVersion() < 20) {
+                biomes.remove(Biome.CHERRY_GROVE);
```
Comment (PierreSchwang) [inline Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java:442]: I'd like a rather more variable approach, as that removes the biome if the datapack is installed as well.
I do think experimental biomes will be part of future releases as well, and we would have to address that for each release

Something like `biomes.removeIf(b -> Registry.BIOME.get(b.getKey()) == null);` which checks if the key is present in the servers registry should do the trick - havent tested that tho

**Review by PierreSchwang — CHANGES_REQUESTED** [review]

File: Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java (hunk 1 above)
Comment (TheMeinerLP) [reply Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java:442]: > I'd like a rather more variable approach, as that removes the biome if the datapack is installed as well.
I asked about that already at paper dc. There said to me. I can create a fake datapack to emulate 1.20. Thats the reason why i do that like so.

Comment (PierreSchwang) [reply Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java:442]: After thinking about that, the whole static block could potentially be simplified to 
```java
BIOMES = Arrays.stream(Biome.values())
                    .filter(b -> Registry.BIOME.get(b.getKey()) == null)
                    .toList();
```
as the remove of the custom biome should be handled by that as well

Comment (TheMeinerLP) [reply Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java:442]: But `Registry.BIOME.get(b.getKey()) == null` is NMS Code. When i am right or ?

Comment (PierreSchwang) [reply Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java:442]: > But `Registry.BIOME.get(b.getKey()) == null` is NMS Code. When i am right or ?

No: https://jd.papermc.io/paper/1.19/org/bukkit/Registry.html#BIOME (does exist in 1.16 as well)

> I asked about that already at paper dc. There said to me. I can create a fake datapack to emulate 1.20. Thats the reason why i do that like so.

You can enable that easily by setting the appropriate value(s) to initial_enabled_packs in server.properties which enables the support for the new biome in 1.19 servers.

Comment (TheMeinerLP) [reply Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java:442]: I will apply that code from you. Thanks.

Comment (PierreSchwang) [reply Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java]: Could you test the simplified version as well? (https://github.com/IntellectualSites/PlotSquared/pull/4016#discussion_r1170636323)

**Comment by PierreSchwang** [comment]
Okay, I've tested quite a bit - seems like the paper api registries are just holders for the enums and do not represent the server registries itself (smh) - so, the cherry groove is always present and my idea won't work. There could be possibly a workaround using the BiomeProvider for all worlds and search for the allowed biomes there - not sure yet. Will take a further look into that tomorrow / later. That is quite a big issue with the (FA)WE BiomeRegistry as well, as we register all biome enum values into the WE Registry - even if they are not accessible (which breaks e.g. `/plot set biome`)
image

**Comment by TheMeinerLP** [comment]
`/plot set biome` works so far with custom biomes and default ones

**Comment by PierreSchwang** [comment]
> `/plot set biome` works so far with custom biomes and default ones

That's not what I meant. The cherry_grove is tap complete in that particular command as well - even if it is not available in the CraftRegistry itself. That's due to the two-staged biome loading functionality of fawes biome registry. We'd need a method to dynamically determine available biomes for both issues. 
For the PS issue I do have an idea already, as the IBukkitAdapter has a method to retrieve registered biomes to the server itself - not sure if that is FAWE specific or exists in regular WE as well.

**Comment by PierreSchwang** [comment]
Okay, I've tested quite a bit with the NMS registries and shit - I'd say for now rollback to your first approach, at least the plugin works again then. But Biomes are still a thing we have to address further in PS and FAWE

**Comment by TheMeinerLP** [comment]
> Okay, I've tested quite a bit with the NMS registries and shit - I'd say for now rollback to your first approach, at least the plugin works again then. But Biomes are still a thing we have to address further in PS and FAWE

I had this stupid idea:
```java
static {
    BIOMES = Arrays.stream(Biome.values())
            .filter(b -> BukkitAdapter.adapt(BiomeType.REGISTRY.get(b.getKey().toString())) != null)
            .toList();
}
```

**Comment by PierreSchwang** [comment]
> > Okay, I've tested quite a bit with the NMS registries and shit - I'd say for now rollback to your first approach, at least the plugin works again then. But Biomes are still a thing we have to address further in PS and FAWE
> 
> I had this stupid idea:
> 
> ```java
> static {
>     BIOMES = Arrays.stream(Biome.values())
>             .filter(b -> BukkitAdapter.adapt(BiomeType.REGISTRY.get(b.getKey().toString())) != null)
>             .toList();
> }
> ```

That sadly won't work as well, as the cherry blossom biome is registered by FAWE as well (which is another issue...)

**Comment by NotMyFault** [comment]
I dropped the last commit, given the change proposed as-is restores the functionality of creating worlds.

**Review by NotMyFault — APPROVED** [review]

**Comment by NotMyFault** [comment]
Maybe someone else from @IntellectualSites/plotsquared-team wants to take a look in the meantime.

File: Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java [hunk 2]
```diff
@@ -434,7 +436,10 @@ private final class BukkitPlotBiomeProvider extends BiomeProvider {
         private static final List<Biome> BIOMES;
 
         static {
-            ArrayList<Biome> biomes = new ArrayList<>(List.of(Biome.values()));
+
+            ArrayList<Biome> biomes = new ArrayList<>(Arrays.stream(Biome.values())
+                    .filter(b -> Registry.BIOME.get(b.getKey()) == null)
```
Comment (SirYwell) [inline Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java:441]:
```suggestion
                    .filter(b -> Registry.BIOME.get(b.getKey()) != null)
```

File: Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java [hunk 3]
```diff
@@ -434,7 +436,10 @@ private final class BukkitPlotBiomeProvider extends BiomeProvider {
...
+                    .toList());
```
Comment (SirYwell) [inline Bukkit/src/main/java/com/plotsquared/bukkit/generator/BukkitPlotGenerator.java:442]: `.collect(Collectors.toCollection(ArrayList::new))`, then the outer ArrayList creation can be dropped

**Review by SirYwell — CHANGES_REQUESTED** [review]

**Comment by NotMyFault** [comment]
@TheMeinerLP Do you want to address the changes requested?

**Comment by NotMyFault** [comment]
@SirYwell might take a look? This PR is needed to unblock v7 on earlier versions.

**Review by SirYwell — APPROVED** [review]
Can't test atm, but looks good code-wise

**Comment by NotMyFault** [comment]
> Can't test atm, but looks good code-wise

I didn't encounter issues with it earlier testing on the FAWE v7 branch with 1.20 on 1.20.

**Pull request merged**